02_bubble/brnbke/main.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(arr []int) []int {
+	changed := true
+	r := make([]int, len(arr))
+	copy(r, arr)
+
+	for {
+		if !changed {
```

Comment:
As Sam says, since this condition is the first thing in the loop and it breaks the loop, you may as well put it in the condition of the loop:
```
for changed {
...
}
```
It does read a bit funny as Go doesn't use "while" for this, opting to use "for" for all loops.

Reply:
Moved the check to the bottom of the look to keep the changed var scoped to the loops, not sure I can implement it as "for changed" as need to reset the changed to false each loop and only exit if not changed

Reply:
You could do:
```
for changed := true; changed; {
    changed = false
    ...
}
```
if you want to keep the var scoped to the loop. But this is fine as is. I won't insist on any further change here.

Reply:
[sorry if you get this twice. I wanted to add it in the review, but the comment has gone from the review.]

You could do:
```
for changed := true; changed; {
    changed = false
    ...
}
```
if you want to keep the var scoped to the loop. But this is fine as is. I won't insist on any further change here.

Reply:
Ah I see what you mean, I assume this is the prefered Go way of doing this. will take note for future, but will leave as is for this to avoid further reviews.

Reply:
I'm not sure if Go-style really has much to say about this, but I prefer to avoid infinite loops with the terminating condition in the body - it's a bit easier to read and know that the loop will actually terminate. In this particular case it's easy enough to tell, but not always.